File: Shundori/attendance.py
"""
Attendance tracking logic.
"""
from datetime import datetime, date
from typing import Optional
from database import get_db, User, AttendanceRecord, Fine, Settings
from utils import get_phnom_penh_now, get_phnom_penh_date, is_before_deadline, is_attendance_window_open
from config import DEFAULT_FINE_AMOUNT
from reports import get_fine_amount


def get_or_create_user(telegram_id: int, username: str = None, full_name: str = None) -> User:
    """Get or create a user in the database."""
    with get_db() as db:
        user = db.query(User).filter(User.telegram_id == telegram_id).first()
        if not user:
            user = User(
                telegram_id=telegram_id,
                username=username,
                full_name=full_name
            )
            db.add(user)
            db.commit()
            db.refresh(user)
        else:
            # Update user info if provided
            updated = False
            if username and user.username != username:
                user.username = username
                updated = True
            if full_name and user.full_name != full_name:
                user.full_name = full_name
                updated = True
            if updated:
                db.commit()
        return user


def record_attendance(telegram_id: int, timestamp: datetime = None, username: str = None, full_name: str = None) -> tuple[bool, str]:
    """
    Record attendance for a user.
    Returns (success, message)
    """
    if timestamp is None:
        timestamp = get_phnom_penh_now()
    
    # Check if window is open
    if not is_attendance_window_open():
        return False, "Attendance window is closed. Please send '1' between 09:00 and 10:00 AM."
    
    current_date = get_phnom_penh_date()
    
    # Get or create user first (outside transaction to avoid nested contexts)
    user = get_or_create_user(telegram_id, username=username, full_name=full_name)
    
    with get_db() as db:
        # Check if already recorded for today
        existing = db.query(AttendanceRecord).filter(
            AttendanceRecord.user_id == user.id,
            AttendanceRecord.date == current_date
        ).first()
        
        if existing:
            return False, "You have already recorded your attendance for today."
        
        # Check if before deadline
        is_on_time = is_before_deadline(timestamp)
        status = 'present' if is_on_time else 'late'
        
        # Create attendance record
        record = AttendanceRecord(
            user_id=user.id,
            date=current_date,
            status=status,
            timestamp=timestamp
        )
        db.add(record)
        
        # If late, create fine
        if not is_on_time:
            fine_amount = get_fine_amount()
            fine = Fine(
                user_id=user.id,
                date=current_date,
                amount=fine_amount
            )
            db.add(fine)
        
        db.commit()
        
        if is_on_time:
            return True, "Good morning! Attendance recorded."
        else:
            return True, "Attendance recorded, but you were late. A fine has been applied."


def process_daily_attendance():
    """
    Process attendance for all members at 10:00 AM.
    Mark absent members and apply fines.
    """
    current_date = get_phnom_penh_date()
    fine_amount = get_fine_amount()
    
    with get_db() as db:
        # Get all active users
        users = db.query(User).filter(User.is_active == True).all()
        
        for user in users:
            # Check if user has attendance record for today
            record = db.query(AttendanceRecord).filter(
                AttendanceRecord.user_id == user.id,
                AttendanceRecord.date == current_date
            ).first()
            
            if not record:
                # Mark as absent
                record = AttendanceRecord(
                    user_id=user.id,
                    date=current_date,
                    status='absent',
                    timestamp=None
                )
                db.add(record)
                
                # Apply fine
                fine = Fine(
                    user_id=user.id,
                    date=current_date,
                    amount=fine_amount
                )
                db.add(fine)
        
        db.commit()


def force_mark_attendance(telegram_id: int, status: str, target_date: date = None) -> bool:
    """
    Force mark attendance for a user (admin only).
    status: 'present' or 'absent'
    """
    if target_date is None:
        target_date = get_phnom_penh_date()
    
    with get_db() as db:
        user = db.query(User).filter(User.telegram_id == telegram_id).first()
        if not user:
            return False
        
        # Remove existing record and fine
        existing_record = db.query(AttendanceRecord).filter(
            AttendanceRecord.user_id == user.id,
            AttendanceRecord.date == target_date
        ).first()
        
        existing_fine = db.query(Fine).filter(
            Fine.user_id == user.id,
            Fine.date == target_date
        ).first()
        
        if existing_record:
            db.delete(existing_record)
        if existing_fine:
            db.delete(existing_fine)
        
        # Create new record
        if status == 'present':
            record = AttendanceRecord(
                user_id=user.id,
                date=target_date,
                status='present',
                timestamp=get_phnom_penh_now()
            )
            db.add(record)
        elif status == 'absent':
            record = AttendanceRecord(
                user_id=user.id,
                date=target_date,
                status='absent',
                timestamp=None
            )
            db.add(record)
            
            # Apply fine
            fine_amount = get_fine_amount()
            fine = Fine(
                user_id=user.id,
                date=target_date,
                amount=fine_amount
            )
            db.add(fine)
        
        db.commit()
        return True

File: Shundori/nodejs/attendance.js
/**
 * Attendance tracking logic
 */
const { pool } = require('./database');
const { getPhnomPenhNow, getPhnomPenhDate, isBeforeDeadline, isAttendanceWindowOpen, config } = require('./config');

async function getOrCreateUser(telegramId, username = null, fullName = null) {
  const client = await pool.connect();
  try {
    let result = await client.query(
      'SELECT * FROM users WHERE telegram_id = $1',
      [telegramId]
    );

    if (result.rows.length === 0) {
      result = await client.query(
        'INSERT INTO users (telegram_id, username, full_name) VALUES ($1, $2, $3) RETURNING *',
        [telegramId, username, fullName]
      );
    } else {
      // Update user info if provided
      if (username || fullName) {
        result = await client.query(
          'UPDATE users SET username = COALESCE($2, username), full_name = COALESCE($3, full_name) WHERE telegram_id = $1 RETURNING *',
          [telegramId, username, fullName]
        );
      }
    }
    return result.rows[0];
  } finally {
    client.release();
  }
}

async function getFineAmount() {
  const client = await pool.connect();
  try {
    const result = await client.query(
      'SELECT value FROM settings WHERE key = $1',
      ['fine_amount']
    );
    if (result.rows.length > 0) {
      return parseFloat(result.rows[0].value);
    }
    return config.DEFAULT_FINE_AMOUNT;
  } finally {
    client.release();
  }
}

async function recordAttendance(telegramId, timestamp = null, username = null, fullName = null) {
  if (!timestamp) {
    timestamp = getPhnomPenhNow().toDate();
  } else if (typeof timestamp === 'number') {
    // Convert Unix timestamp (milliseconds) to Date
    timestamp = new Date(timestamp);
  }

  // Check if window is open
  if (!isAttendanceWindowOpen()) {
    return { success: false, message: 'Attendance window is closed. Please send \'1\' between 09:00 and 10:00 AM.' };
  }

  const currentDate = getPhnomPenhDate();
  const client = await pool.connect();

  try {
    await client.query('BEGIN');

    // Get or create user
    const user = await getOrCreateUser(telegramId, username, fullName);

    // Check if already recorded for today
    const existing = await client.query(
      'SELECT * FROM attendance_records WHERE user_id = $1 AND date = $2',
      [user.id, currentDate]
    );

    if (existing.rows.length > 0) {
      await client.query('ROLLBACK');
      return { success: false, message: 'You have already recorded your attendance for today.' };
    }

    // Check if before deadline
    const isOnTime = isBeforeDeadline(timestamp);
    const status = isOnTime ? 'present' : 'late';

    // Create attendance record
    await client.query(
      'INSERT INTO attendance_records (user_id, date, status, timestamp) VALUES ($1, $2, $3, $4)',
      [user.id, currentDate, status, timestamp]
    );

    // If late, create fine
    if (!isOnTime) {
      const fineAmount = await getFineAmount();
      await client.query(
        'INSERT INTO fines (user_id, date, amount) VALUES ($1, $2, $3)',
        [user.id, currentDate, fineAmount]
      );
    }

    await client.query('COMMIT');

    if (isOnTime) {
      return { success: true, message: 'Good morning! Attendance recorded.' };
    } else {
      return { success: true, message: 'Attendance recorded, but you were late. A fine has been applied.' };
    }
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
}

async function processDailyAttendance() {
  const currentDate = getPhnomPenhDate();
  const fineAmount = await getFineAmount();
  const client = await pool.connect();

  try {
    await client.query('BEGIN');

    // Get all active users
    const users = await client.query('SELECT * FROM users WHERE is_active = TRUE');

    for (const user of users.rows) {
      // Check if user has attendance record for today
      const record = await client.query(
        'SELECT * FROM attendance_records WHERE user_id = $1 AND date = $2',
        [user.id, currentDate]
      );

      if (record.rows.length === 0) {
        // Mark as absent
        await client.query(
          'INSERT INTO attendance_records (user_id, date, status, timestamp) VALUES ($1, $2, $3, NULL)',
          [user.id, currentDate, 'absent']
        );

        // Apply fine
        await client.query(
          'INSERT INTO fines (user_id, date, amount) VALUES ($1, $2, $3)',
          [user.id, currentDate, fineAmount]
        );
      }
    }

    await client.query('COMMIT');
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
}

async function forceMarkAttendance(telegramId, status, targetDate = null) {
  if (!targetDate) {
    targetDate = getPhnomPenhDate();
  }

  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    const userResult = await client.query('SELECT * FROM users WHERE telegram_id = $1', [telegramId]);
    if (userResult.rows.length === 0) {
      await client.query('ROLLBACK');
      return false;
    }

    const user = userResult.rows[0];

    // Remove existing record and fine
    await client.query('DELETE FROM attendance_records WHERE user_id = $1 AND date = $2', [user.id, targetDate]);
    await client.query('DELETE FROM fines WHERE user_id = $1 AND date = $2', [user.id, targetDate]);

    // Create new record
    if (status === 'present') {
      await client.query(
        'INSERT INTO attendance_records (user_id, date, status, timestamp) VALUES ($1, $2, $3, $4)',
        [user.id, targetDate, 'present', getPhnomPenhNow().toDate()]
      );
    } else if (status === 'absent') {
      await client.query(
        'INSERT INTO attendance_records (user_id, date, status, timestamp) VALUES ($1, $2, $3, NULL)',
        [user.id, targetDate, 'absent']
      );

      // Apply fine
      const fineAmount = await getFineAmount();
      await client.query(
        'INSERT INTO fines (user_id, date, amount) VALUES ($1, $2, $3)',
        [user.id, targetDate, fineAmount]
      );
    }

    await client.query('COMMIT');
    return true;
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
}

module.exports = {
  getOrCreateUser,
  recordAttendance,
  processDailyAttendance,
  forceMarkAttendance,
  getFineAmount
};

